apollo-adapters-core/src/commonMain/kotlin/com/apollographql/adapter/core/UnitAdapter.kt

```diff
@@ -0,0 +1,16 @@
+package com.apollographql.adapter.core
+
+import com.apollographql.apollo.api.Adapter
+import com.apollographql.apollo.api.CustomScalarAdapters
+import com.apollographql.apollo.api.json.JsonReader
+import com.apollographql.apollo.api.json.JsonWriter
+
+internal object UnitAdapter : Adapter<Unit> {
+  override fun fromJson(reader: JsonReader, customScalarAdapters: CustomScalarAdapters): Unit {
+    return Unit
+  }
+
+  override fun toJson(writer: JsonWriter, customScalarAdapters: CustomScalarAdapters, value: Unit) {
+    writer.nullValue()
```

Comment:
I think we should just throw here - we can't know what the backend is expecting.

```suggestion
    error("Void used in input position")
```

Reply:
hey @BoD I'm not sure about this change, maybe I'm missing something and/or I'm doing something wrong but I get an error running tests that use the fake resolver:

```
Don't know how to instantiate leaf Void
java.lang.IllegalStateException: Don't know how to instantiate leaf Void
	at com.apollographql.apollo.api.DefaultFakeResolver.resolveLeaf(fakeResolver.kt:280)
```

Then I add the mapping from scalar "Void" to Kotlin `Unit`
```kotlin
    override fun resolveLeaf(context: FakeResolverContext): Any =
        when (context.mergedField.type.rawType().name) {
            "AnyType" -> "String"
            "Void" -> Unit
            else -> super.resolveLeaf(context)
        }
```
and I end up in the `error()` case:

```
Void used in input position
java.lang.IllegalStateException: Void used in input position
	at com.veeva.link.library.network.UnitAdapter.toJson(UnitAdapter.kt:14)
	at com.veeva.link.library.network.UnitAdapter.toJson(UnitAdapter.kt:8)
```

The issue happens in this line:
```kotlin
private val data = MyMutation.Data(MyFakeResolver()) {}
```

Do you think I'm doing something wrong in my tests?

With `writer.nullValue()` everything seems fine. Would it be fine to just keep it?

Reply:
Hmm I'm not 100% sure of what's happening here. Is `Void` used in arguments of your operations or in input types? Can you share more of the stacktrace maybe.

> With writer.nullValue() everything seems fine. Would it be fine to just keep it?

It would make more sense to throw as this really shouldn't be used in input positions.

Reply:
These are all the occurrences of `Void` in the schema:

```graphql
"""
Root mutation
"""
type MyMutation {
  userDetails(data: UpdateUserDetailsInput!): Void @deprecated(reason: "This mutation is redundant. Use UserSettings instead!")

  userSettings(data: UpdateUserSettingsInput!): Void

  termsAndConditions(data: UpdateTermsAndConditionsInput!): Void

  setPassword(data: SetUserPasswordInput!): Boolean!

  changePassword(data: ChangeUserPasswordInput!): Boolean!

  uploadImage(image: Upload!): ImageUploadResult!

  deleteImage: Void
}

"""
Represents NULL values
"""
scalar Void
```

Full stacktrace:

```
Void used in input position
java.lang.IllegalStateException: Void used in input position
	at com.library.network.UnitAdapter.toJson(UnitAdapter.kt:14)
	at com.library.network.UnitAdapter.toJson(UnitAdapter.kt:8)
	at com.apollographql.apollo.api.ObjectBuilderKt.adaptValue(ObjectBuilder.kt:61)
	at com.apollographql.apollo.api.FakeResolverKt.buildFieldOfNonNullType(fakeResolver.kt:235)
	at com.apollographql.apollo.api.FakeResolverKt.buildFieldOfType(fakeResolver.kt:161)
	at com.apollographql.apollo.api.FakeResolverKt.buildFieldOfType(fakeResolver.kt:156)
	at com.apollographql.apollo.api.FakeResolverKt.buildFieldOfNonNullType(fakeResolver.kt:203)
	at com.apollographql.apollo.api.FakeResolverKt.buildFieldOfType(fakeResolver.kt:161)
	at com.apollographql.apollo.api.FakeResolverKt.buildFakeObject(fakeResolver.kt:113)
	at com.apollographql.apollo.api.FakeResolverKt.buildData(fakeResolver.kt:338)
	at com.apollographql.apollo.api.FakeResolverKt.buildData(fakeResolver.kt:353)
	at com.library.network.api.graphql.lama.TermsAndConditionsMutation$Companion.Data(TermsAndConditionsMutation.kt:83)
	at com.feature.login.interactor.UpdateTermsAndConditionsInteractorImplTest.<init>(UpdateTermsAndConditionsInteractorImplTest.kt:28)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
	at org.junit.runners.BlockJUnit4ClassRunner.createTest(BlockJUnit4ClassRunner.java:250)
	at org.junit.runners.BlockJUnit4ClassRunner.createTest(BlockJUnit4ClassRunner.java:260)
	at org.junit.runners.BlockJUnit4ClassRunner$2.runReflectiveCall(BlockJUnit4ClassRunner.java:309)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.BlockJUnit4ClassRunner.methodBlock(BlockJUnit4ClassRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:112)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:40)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:60)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:52)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy1/jdk.proxy1.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

Reply:
Thanks! Can you try to remove your case for `Void` in `resolveLeaf` and instead add a case for it and return true in `resolveMaybeNull`. 

That way, `Void` will be resolved as `null`, and the adapter won't be called.

Reply:
yeah that did the trick, thanks!

```kotlin
    override fun resolveMaybeNull(context: FakeResolverContext): Boolean =
        if (context.mergedField.type.rawType().name == "Void") {
            true
        } else {
            super.resolveMaybeNull(context)
        }
```